model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto

```diff
@@ -1063,6 +1087,10 @@ message StateAppendRequest {
   // Represents a part of a logical byte stream. Elements within
   // the logical byte stream are encoded in the nested context and
   // multiple append requests are concatenated together.
+  // For OrderedListState, elements of TimeStampedValue<T> should be encoded
+  // with TimestampedValueCoder.of(LengthPrefixCoder.of(Coder<T>)), so that
+  // the request handler knows how to decode timestamps from the data without
```

Comment:
nit: we refer to this as "sort key" above. Maybe just mention that sort keys should be encoded as timestamps?

Reply:
Looks like TimestampedValueCoder is not well defined, and +1 to references to keys, not timestamps. 

Should we use `KV<VarInt, LP<value>>`? Or is it worth defining a bigendian one? How do we want to treat negatives?

Reply:
> Looks like TimestampedValueCoder is not well defined

TimestampedValueCoder is defined as an inner class in `TimeStampedValue` 
https://github.com/apache/beam/blob/413af1289373079319ba77b3f233751fbdfc3bc9/sdks/java/core/src/main/java/org/apache/beam/sdk/values/TimestampedValue.java#L88
. It encodes the value first and then the timestamp.

> Should we use KV<VarInt, LP<value>>? Or is it worth defining a bigendian one? How do we want to treat negatives?

I use TimeStampedValue<T> in the proto and implementation instead of a KV<,> because it is originally used in the OrderedListState interface https://github.com/apache/beam/blob/f0d06051072da5b64c496dc8456e56bab9aca71b/sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java#L29. 
Another benefit of using TimeStampedValue<T> is that we can reuse of all the existing prefetched mechanisms without reimplementing them. 

Here is a [long discussion thread](https://docs.google.com/document/d/1U77sAvE6Iy9XsVruRYHxPdFji7nqS6HPi1XU8fhyrxs/edit?disco=AAAAHwml1Yg) of why we choose to use timestamp as the sort key in the original design doc. As Reuven pointed out, timestamp is the most common sort key and we can easily converted it from and to an int64. 

```orderedList.add(TimestampedValue.of(element, Instant.of(sequence number)));```

This is valid for both positives and negatives (which will be represented as timestamps prior to epoch 1970-01-01T00:00:00Z).

Reply:
TimestampedValueCoder is not a well defined portable/cross-language Beam coder. We can continue to use TimestampedValue in the Java code if it makes life easier, but IMHO we should be using a better defined coder. `KV<VarInt64, LP<value>>` seems like the most natural option here. (Another bonus: we don't have to dither about endianness or millis vs micros, even if it "usually" holds a timestamp.)

Reply:
This sounds a bit not counter-intuitive, but TimestampedValue is not equivalent to KV<VarInt64, LP<value>>. Particularly, TimestampedValueCoder encodes/and decodes the value first and then timestamp.
https://github.com/apache/beam/blob/413af1289373079319ba77b3f233751fbdfc3bc9/sdks/java/core/src/main/java/org/apache/beam/sdk/values/TimestampedValue.java#L110-L111

But KV<VarInt64, LP<value>> does that in an opposite order. In order to maintain the compatibility of these two coders and makes the code easier, how about we use coder of KV<LP<value>, VarInt64>> instead?

@robertwb

Reply:
Given these are sort _keys_, let's keep them in the key slot. (In other words, I'd rather optimized for cleanliness of the spec over matching an existing internal implementation detail.)

It does not matter that the `TimestampedValue$TimestampedValueCoder` does not agree here, we would just use a different `Coder<TimestampedValue>`. (We need to explicitly ensure that the coder for T is LP anyway, so we shouldn't be inferring coders here.)

Reply:
Sounds reasonable. I will go ahead to change the implementation I have right now to match this change once this proto change is pushed. Thanks!